[ResponseOps] Copy dataViewSpecSchema to rules params package

guskovaue opened the pull request:
Fixed: https://github.com/elastic/kibana/issues/207832

<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!

guskovaue commented:
/ci

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

js-jankisalvi commented on src/platform/packages/shared/response-ops/rule_params/common/data_view_spec_schema/data_view_spec_schema.ts:
```diff
@@ -0,0 +1,50 @@
+
+import { schema } from '@kbn/config-schema';
+import { MAX_DATA_VIEW_FIELD_DESCRIPTION_LENGTH } from './constants';
+import { fieldSpecSchema, runtimeFieldSchema, serializedFieldFormatSchema } from './schemas';
```
super nit: if these schemas are going to be used only in `dataViewSpecSchema `. They could be defined here instead of in a separate file.

js-jankisalvi commented:
> Shouldn't we have this folder follow the same logic as the other schemas in `rule_params`?

I assumed this is in common folder as it will be part of multiple rule params. Is it not the case here?

> We are exporting a few things that arent used anywhere atm. We need to remember to doublecheck whether all these exports are used in the follow-up PR.

agree

guskovaue commented:
@adcoelho & @js-jankisalvi 
>  Shouldn't we have this folder follow the same logic as the other schemas in rule_params
This schema needs for internal use only as I understood. For for rule threshold and for o11y for metric_inventory_threshold and metric_threshold inside our package rules-params package (Am I right @benakansara ?) That's why it's in common.

> Instead of creating a folder data_view_spec_schema, it would be nice to have a file common/data_view_spec_schema.ts
I like it so I changed it.

> We are exporting a few things that arent used anywhere atm. We need to remember to doublecheck whether all these > exports are used in the follow-up PR.
Good point. I deleted all exports for now. If @benakansara will need something to export, better to add it lately.

> Also, I find it a bit confusing to have this PR in isolation.
I was part of another PR, but it's big and probably needs more time for review. So I want to merge it to unblock us and other team asap.

guskovaue commented:
/ci

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/269839)
* Commit: ebd5560dbb21ef64ba3dff97d6de19f16628a102

### Failed CI Steps
* [Build Kibana Distribution](https://buildkite.com/elastic/kibana-pull-request/builds/269839#019494ee-83c8-4913-8ef9-304a967af9ca)

### History
* :yellow_heart: [Build #269360](https://buildkite.com/elastic/kibana-pull-request/builds/269360) was flaky 0e5d56f9e66e51f387dfc76f3af466f1095aa546

cc @guskovaue

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269839","number":269839,"commit":"ebd5560dbb21ef64ba3dff97d6de19f16628a102"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269360","number":269360,"commit":"0e5d56f9e66e51f387dfc76f3af466f1095aa546"}],"number":269839}
buildkite-pr-comment-->

adcoelho commented on src/platform/packages/shared/response-ops/rule_params/common/data_view_spec_schema.ts:
```diff
@@ -0,0 +1,184 @@
+
+const runtimeFieldNonCompositeFieldsSpecTypeSchema = schema.oneOf(
+  RUNTIME_FIELD_TYPES2.map((runtimeFieldType) => schema.literal(runtimeFieldType)) as [
+    Type<RuntimeType>
```
This is `RUNTIME_FIELD_TYPES2` but uses `type RuntimeType = (typeof RUNTIME_FIELD_TYPES)[number];`.

Shouldn't we instead use/create `type RuntimeType2 = (typeof RUNTIME_FIELD_TYPES2)[number];`?

guskovaue replied:
Good point. I've raised it with the owner of the code.

guskovaue replied:
@adcoelho Just received the answer from data view team:
I don’t think anyone on our team knows the full context here since it’s from before we owned data views. but looking at the code, it seems like RUNTIME_FIELD_TYPES2 is a subset of RUNTIME_FIELD_TYPES, excluding composite since there’s a separate schema for validating composite runtime fields. every value from RUNTIME_FIELD_TYPES2 is a valid RuntimeType, but I see how the casting to RuntimeType here results in a wider type than is actually valid. 

A separate RuntimeType2 type could be created here to correctly omit composite, but I’m not sure if that would cascade into a bunch of type failures or not. unless it’s causing specific issues for you, I probably wouldn’t worry much about it considering it’s used for API payload validation, and the validation still works correctly.

guskovaue replied:
If they do not have any planes to fix it, I think we can just continue with the exact copy of their code since it's working. Our code is just a shadow of their code. At least intention was like this. 
WDYT?

adcoelho replied:
In this case, it wouldn't cascade into multiple type failures since only we use this schema :D but I'm ok with leaving it like this to ensure the schemas don't start already diverging.

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/269896)
* Commit: 3f2b37ce0f3c08b9ecd1ab6cfc37f49bf5305f3d

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #269839](https://buildkite.com/elastic/kibana-pull-request/builds/269839) failed ebd5560dbb21ef64ba3dff97d6de19f16628a102
* :yellow_heart: [Build #269360](https://buildkite.com/elastic/kibana-pull-request/builds/269360) was flaky 0e5d56f9e66e51f387dfc76f3af466f1095aa546

cc @guskovaue

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269896","number":269896,"commit":"3f2b37ce0f3c08b9ecd1ab6cfc37f49bf5305f3d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269839","number":269839,"commit":"ebd5560dbb21ef64ba3dff97d6de19f16628a102"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269360","number":269360,"commit":"0e5d56f9e66e51f387dfc76f3af466f1095aa546"}],"number":269896}
buildkite-pr-comment-->

benakansara commented:
> This schema needs for internal use only as I understood. For for rule threshold and for o11y for metric_inventory_threshold and metric_threshold inside our package rules-params package (Am I right @benakansara ?) That's why it's in common.

right, `dataViewSpecSchema` will be used by multiple rules and within `rule_params` package

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/270325)
* Commit: f9924a0582308a2f262bfc3c6366b127095902bd

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #269896](https://buildkite.com/elastic/kibana-pull-request/builds/269896) succeeded 3f2b37ce0f3c08b9ecd1ab6cfc37f49bf5305f3d
* :broken_heart: [Build #269839](https://buildkite.com/elastic/kibana-pull-request/builds/269839) failed ebd5560dbb21ef64ba3dff97d6de19f16628a102
* :yellow_heart: [Build #269360](https://buildkite.com/elastic/kibana-pull-request/builds/269360) was flaky 0e5d56f9e66e51f387dfc76f3af466f1095aa546

cc @guskovaue

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/270325","number":270325,"commit":"f9924a0582308a2f262bfc3c6366b127095902bd"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269896","number":269896,"commit":"3f2b37ce0f3c08b9ecd1ab6cfc37f49bf5305f3d"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269839","number":269839,"commit":"ebd5560dbb21ef64ba3dff97d6de19f16628a102"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/269360","number":269360,"commit":"0e5d56f9e66e51f387dfc76f3af466f1095aa546"}],"number":270325}
buildkite-pr-comment-->

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/208353">](https://github.com/elastic/kibana/pull/208353)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
